src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -271,34 +278,40 @@ DatabaseShardImp::prepareShard(std::uint32_t shardIndex)
             std::to_string(earliestShardIndex()));
     }
 
-    // If we are synced to the network, check if the shard index
-    // is greater or equal to the current shard.
-    auto seqCheck = [&](std::uint32_t seq) {
+    // If we are synced to the network, check if the shard
+    // index is greater or equal to the current shard.
+    auto seqCheck = [&](std::uint32_t ledgerSeq) {
         // seq will be greater than zero if valid
-        if (seq >= earliestLedgerSeq() && shardIndex >= seqToShardIndex(seq))
-            return fail("has an invalid index");
+        if (ledgerSeq >= earliestLedgerSeq() &&
+            shardIndex >= seqToShardIndex(ledgerSeq))
+        {
+            return fail("invalid index");
+        }
         return true;
     };
     if (!seqCheck(app_.getLedgerMaster().getValidLedgerIndex() + 1) ||
         !seqCheck(app_.getLedgerMaster().getCurrentLedgerIndex()))
     {
-        return false;
+        return fail("invalid index");
     }
 
-    if (shards_.find(shardIndex) != shards_.end())
-    {
-        JLOG(j_.debug()) << "shard " << shardIndex
-                         << " is already stored or queued for import";
-        return false;
-    }
+    std::lock_guard lock(mutex_);
+    assert(init_);
+
+    if (!canAdd_)
+        return fail("cannot be stored at this time");
 
     // Check available storage space
     if (fileSz_ + avgShardFileSz_ > maxFileSz_)
```

Comment:
Maybe this fix doesn't go into this PR as this is a preexisting issue: this method only checks if we have enough space to add one additional shard. If there are 10GB available to the shards directory (`DatabaseShardImp::dir_`), and the `ShardArchiveHandler` receives a request to download ten shards, each download will pass this storage space check, but there won't be enough space to hold all ten shards.